README.md

```diff
@@ -1 +1,108 @@
-# runner
\ No newline at end of file
+# runner design doc
+
+## Background
+This repo is intended to hold the client, server (api), and underlying (library) golang code for a gRPC-based runner of linux commands. 
+
+(phew ... that's a mouthful)
+
+Not only that, but it should be secured by TLS, honor cgroups and, optionally, linux namespaces, and stream data back to the client from the running process. 
+Impossible you say?! Maybe. Think of it as making docker runtime without the image management.
+
+This document will outline my approach to make this a reality. 
+
+### Broad strokes
+
+I hope to implement this in a series of PR's that follow this schedule:
+
+0. this doc
+1. create the library for the process runner and cgroup/ns support (perhaps based on client id roles)
+2. basic TLS-secured gRPC client-server chat. The API will wrapper the process runner library
+3. support for hard coded gRPC streaming data. Something simple like a clock tick
+4. Spawn some basic jobs, capture stderr/stdout and pump them into the streams
+5. Work on job control. Cancelling a process, ensuring the channels/goroutines/resources clean up correctly, etc. 
+6. Add cgroups and resource isolation 
+
+(this may change)
+
+Stretch goals
+
+7. some key tests
+8. docker packaging
+9. a nice client
+10. magic
+
+## API
+
+Exposed gRPC "Runner" service methods:
+```
+Run(ClientID, cmd, args...) (Process, error)
+Stdout(ClientID, Process) (Stream, error)
+Status(ClientID, Process) (Status, error)
+Abort(ClientID, Process) (Status, error)
+```
+
+where: 
+ * the `Run` method that can launch new processes for authenticated callers.
+   * the "role" of the authenticated client will determine the cgroup style the process runs in. This will be extracted from client mTLS cert.
+   * `Run` takes `cmd` and `args` args for the intended linux command and related cmdline parameters and returns a `Process` message
+ * `Process` is an abstracted process ID for the process control calls: `Stdout`, `Status`, and `Abort`. It will be a UUID that maps to the underlying os pid. 
+ * `Status` will return an enum of `COMPLETED|RUNNING|ERROR|ABORTED` and, ideally, the exit code
+
+## AuthN / AuthZ
+
+mTLS client certs will be used for authN. The client cert will map to a set of "roles" that determine the class of cgroup processes will run under. 
```

Comment:
It looks like these roles determine what level of cgroups are afforded to the user, but this is different from AuthZ. Can you add details for a simple Authorization scheme? You could reuse these cgroup roles for it, or handle it separately.

Reply:
will do ... thanks